drivers/entropy/entropy_esp32.c

```diff
@@ -8,52 +8,77 @@
 
 #include <string.h>
 #include <drivers/entropy.h>
+#include <soc/dport_reg.h>
+#include <soc/rtc.h>
+#include <soc/wdev_reg.h>
+#include <soc/rtc_cntl_reg.h>
+#include <soc/apb_ctrl_reg.h>
+#include <esp_system.h>
+#include <soc.h>
+
+extern int esp_clk_cpu_freq(void);
+extern int esp_clk_apb_freq(void);
 
 static inline uint32_t entropy_esp32_get_u32(void)
 {
-	/*
-	 * APB Address:   0x60035144 (Safe,slower writes)
-	 * DPORT Address: 0x3ff75144 (write bug, fast writes)
-	 * In this case it won't make a difference because it is read only
-	 * More info available at:
-	 * https://www.esp32.com/viewtopic.php?f=2&t=3033&p=14227
-	 * also check: ECO and Workarounds for Bugs Document, point 3.3
+	/* The PRNG which implements WDEV_RANDOM register gets 2 bits
+	 * of extra entropy from a hardware randomness source every APB clock cycle
+	 * (provided WiFi or BT are enabled). To make sure entropy is not drained
+	 * faster than it is added, this function needs to wait for at least 16 APB
+	 * clock cycles after reading previous word. This implementation may actually
+	 * wait a bit longer due to extra time spent in arithmetic and branch statements.
+	 *
+	 * As a (probably unncessary) precaution to avoid returning the
+	 * RNG state as-is, the result is XORed with additional
+	 * WDEV_RND_REG reads while waiting.
 	 */
-	volatile uint32_t *rng_data_reg = (uint32_t *)DT_INST_REG_ADDR(0);
 
-	/* Read just once.  This is not optimal as the generator has
-	 * limited throughput due to scarce sources of entropy, specially
-	 * with the radios turned off.  Might want to revisit this.
+	/* This code does not run in a critical section, so CPU frequency switch may
+	 * happens while this code runs (this will not happen in the current
+	 * implementation, but possible in the future). However if that happens,
+	 * the number of cycles spent on frequency switching will certainly be more
+	 * than the number of cycles we need to wait here.
 	 */
-	return *rng_data_reg;
+	uint32_t cpu_to_apb_freq_ratio =
+		esp_clk_cpu_freq() / esp_clk_apb_freq();
+
+	static uint32_t last_ccount;
+	uint32_t ccount;
+	uint32_t result = 0;
+
+	do {
+		ccount = XTHAL_GET_CCOUNT();
+		result ^= REG_READ(WDEV_RND_REG);
+	} while (ccount - last_ccount < cpu_to_apb_freq_ratio * 16);
+	last_ccount = ccount;
+	return result ^ REG_READ(WDEV_RND_REG);
 }
 
-static int entropy_esp32_get_entropy(const struct device *device,
-				     uint8_t *buf, uint16_t len)
+static int entropy_esp32_get_entropy(const struct device *device, uint8_t *buf,
+				     uint16_t len)
 {
-	while (len) {
-		uint32_t v = entropy_esp32_get_u32();
-
-		if (len >= sizeof(v)) {
-			memcpy(buf, &v, sizeof(v));
-
-			buf += sizeof(v);
-			len -= sizeof(v);
-		} else {
-			memcpy(buf, &v, len);
-			break;
-		}
+	assert(buf != NULL);
+	uint8_t *buf_bytes = (uint8_t *)buf;
+
+	while (len > 0) {
+		uint32_t word = entropy_esp32_get_u32();
+		uint32_t to_copy = MIN(sizeof(word), len);
+
+		memcpy(buf_bytes, &word, to_copy);
```

Comment:
You can use `memmove` here since `buf_bytes` and `word` can't overlap.  It might be faster (assuming the compiler doesn't detect they're disjoint).